### PR TITLE
[4.0] Pagination seo friendly

### PR DESCRIPTION
**Pull request by rajkumaar23**
Pull Request for Issue #26881 

### Summary of Changes
- Appends a `limitstart` param to the action on pagination links

### Testing Instructions
- Ensure the pagination doesn't break with this PR.
- Verify if `limitstart` param is appended & modified while paginating.
- Check if proper page is rendered when `limitstart` is modified in address bar.

### Documentation Changes Required
NIL

### CODE REVIEW

**Comment by Quy** [comment]
Please revert coding style changes per [Joomla! Coding Standards](https://developer.joomla.org/coding-standards/html.html).

**Comment by rajkumaar23** [comment]
> Please revert coding style changes per [Joomla! Coding Standards](https://developer.joomla.org/coding-standards/html.html).

Okay sure, Can you let me know how to populate the pages that use pagination? Because I don’t seem to have enough data to paginate.

**Comment by Quy** [comment]
Try installing `Testing Sample Data`. Then do a search on the frontend.

**Comment by rajkumaar23** [comment]
> Try installing `Testing Sample Data`. Then do a search on the frontend.

Thanks @Quy ! I was able tot test with the dummy data and I've reverted as per coding standards and fixed another bug in the PR too.

**Comment by rajkumaar23** [comment]
Thanks @Quy ! I was able to test with the dummy data and I've reverted as per coding standards and fixed another bug in the PR too.

**Comment by rajkumaar23** [comment]
@Quy Could you please review this PR ?

**Comment by ChristineWk** [comment]
I have tested this item :red_circle: unsuccessfully on 81a7296c146a10982e01d857470335ea61c2771b<br /><br />When going to i.e. page 2 (on pagination link) I get error 404.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

**Comment by rajkumaar23** [comment]
@ChristineWk Can you please check now ?

**Comment by ChristineWk** [comment]
I have tested this item :red_circle: unsuccessfully on e3a52e5514ca7dba479c17a653905932fee700af<br /><br />Error 404 has gone. On 2nd page > in address bar I see: blog?limitstart=5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on e3a52e5514ca7dba479c17a653905932fee700af<br /><br />Error 404 has gone. On 2nd page > in address bar I see: blog?limitstart=5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27489">issues.joomla.org/tracker/joomla-cms/27489</a>.</sub>

File: layouts/joomla/pagination/link.php [hunk 1]
```diff
@@ -54,6 +54,17 @@
 
 if ($displayData['active'])
 {
+	$uriUpdatingScript = "var re = new RegExp('([?&])' + 'limitstart' + '=.*?(&|$)', 'i');
+						  var uri = document.adminForm.action;
+    	   				  var separator = uri.indexOf('?') !== -1 ? '&' : '?';
```
Comment (Quy) [inline layouts/joomla/pagination/link.php]: Please use tabs and not spaces for indentation.

Comment (rajkumaar23) [reply layouts/joomla/pagination/link.php]: Okay, sure.

**Comment by Hackwar** [comment]
I agree that the initial issue is a bug, but I don't understand why the method from J3 isn't still used. Can someone please explain why we should do it this way?

**Comment by Fedik** [comment]
I am sorry to say, but the fix are wrong, such JavaScript hacking not acceptable

**Comment by rajkumaar23** [comment]
> I am sorry to say, but the fix are wrong, such JavaScript hacking not acceptable

Since the pagination itself is done via Javascript, I couldn’t think of any other way to update the URL with the limitstart value. Could you suggest a better idea?

**Comment by rajkumaar23** [comment]
> I agree that the initial issue is a bug, but I don't understand why the method from J3 isn't still used. Can someone please explain why we should do it this way?

Can you please let me know how it was done in J3 ? I’m new to Joomla! and I checked the branches here, and everywhere it’s done in JavaScript only.

File: layouts/joomla/pagination/link.php (hunk 1 above)
Comment (rajkumaar23) [reply layouts/joomla/pagination/link.php]: @Quy I have updated.

**Comment by SharkyKZ** [comment]
In J3 frontend template had an override without JS in it.

Maybe in 4.0 we could do the other way around, make it JS-free by default and add an override for backend template instead.

**Comment by rajkumaar23** [comment]
I guess then the complete pagination code has to change if it has to be no-JS. All I’ve done is append the `limitstart` param to the form action, added to the already existing JS code that modifies the hidden `limitstart` field in the form and triggers it. As far as this issue is concerned, the task was to make different URLs render different content.

**Comment by Fedik** [comment]
You put all your "multi-line" javascript code in to oneclick attribute, do not do this ever.
Also please check how the `form` element works, you no need javascript to append a value that already there (use appropriate submit method GET/POST)

About correct fix, the pagination need to be a list of links, for frontend.
That @SharkyKZ already said

**Comment by rajkumaar23** [comment]
@Fedik I'm sorry, but I couldn't understand what you're trying to say. If I'm not wrong, do you mean all this pagination forms has to be GET, so the URL will have `limitstart` appended automatically ?

**Comment by Fedik** [comment]
Pagination should be a list of links, without javascript, without form.

**Comment by rajkumaar23** [comment]
> Pagination should be a list of links, without javascript, without form.

@Fedik Okay, so you mean to say the current implementation for pagination has to be changed to links. I thought that was out of scope for the related issue. I'll check how to do it.

**Comment by Fedik** [comment]
Check @SharkyKZ comment, it should be changed only for the site, and stay unchanged for administrator.

**Comment by rajkumaar23** [comment]
> Check @SharkyKZ comment, it should be changed only for the site, and stay unchanged for administrator.

Sorry! I'm new to Joomla! Can I get some guidance on what's the `override` he was talking about ?

**Comment by Fedik** [comment]
I think these 2 links

https://docs.joomla.org/Layout_Overrides_in_Joomla#JLayout_Micro-Layout_Overrides
https://docs.joomla.org/J3.x:Sharing_layouts_across_views_or_extensions_with_JLayout

**Comment by SharkyKZ** [comment]
Templates support layout overrides. You can copy `layouts/joomla/pagination/link.php` to `templates/cassiopeia/html/layouts/joomla/pagination/link.php` to create an override for frontend template and remove JS code there.

Or the other way around, copy the file as-is to backend template: `administrator/templates/atum/html/layouts/joomla/pagination/link.php` and then edit original file to remove JS code.

There are other alternatives to consider like creating multiple standard layouts and making them configurable in pagination calls.

**Comment by rajkumaar23** [comment]
@SharkyKZ Can you please check my latest commit if I've correctly understood what you told ?

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	$class = 'active';
+	$onClick = 'document.adminForm.' . $item->prefix . $limit . '; Joomla.submitform();return false;';
+}
+else
+{
+	$class = (property_exists($item, 'active') && $item->active) ? 'active' : 'disabled';
+}
+
+// Retrieving the URL for anchor with limitstart param added/modified
+$currentURL = "http://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
```
Comment (zero-24) [inline templates/cassiopeia/html/layouts/joomla/pagination/link.php]: hmm we should not hardcode http here we might should also better use `JUri` / `\Joomla\CMS\Uri\Uri`

**Review by zero-24 — CHANGES_REQUESTED** [review]

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php (hunk 2 above)
Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: Sure, thanks.

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: Is it fine now?

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	parse_str($urlParts['query'], $params);
+}
+else
+{
+	$params = array();
+}
+
+$params['limitstart'] = $item->base ?: 0;
+$urlParts['query'] = http_build_query($params);
+$updatedURL = http_build_url($urlParts);
```
Comment (Fedik) [inline templates/cassiopeia/html/layouts/joomla/pagination/link.php]: why you doing this?

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: That is the href I've set for the pagination button. How else would I navigate to the next `page` without `limitstart` ?

Comment (Fedik) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: use `href="<?php echo $item->link; ?>"` it already contain all you need

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: I did var_dump on `$item->link` when I was on `/article-category-list` page, but the value was not current URL.

Comment (Fedik) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: what URL was in your var_dump?
the `$item->link` contain a correct link for every pagination item, you no need to modify it, it already prepared in `Pagination` class

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: I think I did a var_dump($item) in administrator page, and received different `link`. Thanks, will use that here.

**Comment by rajkumaar23** [comment]
@Fedik I think now everything is sorted.

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php (hunk 3 above)
Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: Sorry, that was my mistake. I have used it now, thanks :)

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+if ($icon !== null)
+{
+	$display = '<span class="' . $icon . '" aria-hidden="true"></span>';
+}
+
+if ($displayData['active'])
+{
+
+	if ($item->base > 0)
```
Comment (Fedik) [inline templates/cassiopeia/html/layouts/joomla/pagination/link.php]: you can remove this section, it was need for javascript

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		$limit = 'limitstart.value=' . $item->base;
+	}
+	else
+	{
+		$limit = 'limitstart.value=0';
+	}
+
+	$class = 'active';
+	$onClick = 'document.adminForm.' . $item->prefix . $limit . '; Joomla.submitform();return false;';
```
Comment (Fedik) [inline templates/cassiopeia/html/layouts/joomla/pagination/link.php]: and this also not need anymore

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: Done :+1: Removed.

**Comment by Fedik** [comment]
Now it looks much better than first version of it :smiley: 
Left only a testing, I guess.

**Comment by rajkumaar23** [comment]
> Now it looks much better than first version of it smiley
> Left only a testing, I guess.

Haha. This is my first PR with so many comments. OpenSource contributions are indeed best in learning. Thanks! 🔥️

File: templates/cassiopeia/html/layouts/joomla/pagination/link.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+switch ((string) $item->text)
+{
+	// Check for "Start" item
+	case Text::_('JLIB_HTML_START') :
+		$icon = 'fa fa-angle-double-left';
+		$aria = Text::sprintf('JLIB_HTML_GOTO_POSITION', strtolower($item->text));
+		break;
+
+	// Check for "Prev" item
+	case $item->text === Text::_('JPREV') :
```
Comment (Quy) [inline templates/cassiopeia/html/layouts/joomla/pagination/link.php]:
```suggestion
	case Text::_('JPREV') :
```

Comment (rajkumaar23) [reply templates/cassiopeia/html/layouts/joomla/pagination/link.php]: Done.

**Comment by brianteeman** [comment]
I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong

**Comment by rajkumaar23** [comment]
> I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong

Alright then, does having the start param in URL on administrator routes cause any issue? If not, I don’t think this override is required. I can edit the original link.php, but as I was told in this thread that only frontend needs to have that param in the URL, override was the only possible way

**Comment by brianteeman** [comment]
Why does it have to be the same layout for both site and admin?

**Comment by rajkumaar23** [comment]
I didn’t get you @brianteeman :(

**Comment by Fedik** [comment]
@brianteeman there was comment @SharkyKZ suggested to make "this" as default, and then make override for admin template (should be simple), or make them configurable in pagination calls to use different layout depend from needs (this more advanced topic).

Admin use different pagination to keep adminForm "state". The form has extra fields, like CSR token, option, view etc, and can have more depend from component. As far as I know.

**Comment by brianteeman** [comment]
My view is that it should be
> or make them configurable in pagination calls to use different layout depend from needs

**Comment by rajkumaar23** [comment]
@Fedik What do you think could I possibly do to fix this ?

**Comment by Fedik** [comment]
It is good as it is for now, it fixes the issue.

What was suggested is much complex work to do, and can be done any time later.

**Comment by rajkumaar23** [comment]
> It is good as it is for now, it fixes the issue.
> 
> What was suggested is much complex work to do, and can be done any time later.

Can someone test this ?

**Comment by brianteeman** [comment]
Unless I am very much mistaken it is not complex at all
See #27516 for a very quick example

**Comment by Fedik** [comment]
I think we have a different look on what is "configurable pagination" :wink: 

With your suggestion it still not possible to use simple links in backend, or javascript version on front end. It hardly can call "configurable".

The Pagination class allow to specify the layout, that allow to have 2 (an more) different layout at same time
https://github.com/joomla/joomla-cms/blob/4779815c0346c1ba273a6cac4b6d346771f1c8ba/libraries/src/Pagination/Pagination.php#L436-L442

A complexity here is to check all call of `$pagination->getPaginationLinks()` (both admin/site side) and adjust it to use specific layout, and make sure all works.

**Comment by brianteeman** [comment]
> I think we have a different look on what is "configurable pagination" 😉

Maybe so but my PR resolves the issue in the original report without creating template overrides

**Comment by Fedik** [comment]
yeap, that also ok, I don't mind

**Comment by brianteeman** [comment]
as i said before 
> I am coming to this late but to me it is crazy that we need to create an override. If we can't make the core work with the default templates that it ships with there is something very wrong

**Pull request closed without merging**